src/work.metal

```diff
@@ -0,0 +1,127 @@
+        vv[d / 2].x, vv[d / 2].y)
+
+#define G2v_split(m0, m1, m2, m3, a, vb1, vb2, c, vd1, vd2) \
+    G32(m0, m1, m2, m3, vv[a / 2], vb1, vb2, vv[c / 2], vd1, vd2)
+
+#define ROUND(m0, m1, m2, m3, m4, m5, m6, m7, m8, m9, m10, m11, m12, m13, m14, \
+              m15)                                                             \
+    do {                                                                       \
+        G2v(m0, m1, m2, m3, 0, 4, 8, 12);                                      \
+        G2v(m4, m5, m6, m7, 2, 6, 10, 14);                                     \
+        G2v_split(m8, m9, m10, m11, 0, vv[5 / 2].y, vv[6 / 2].x, 10,           \
+                  vv[15 / 2].y, vv[12 / 2].x);                                 \
+        G2v_split(m12, m13, m14, m15, 2, vv[7 / 2].y, vv[4 / 2].x, 8,          \
+                  vv[13 / 2].y, vv[14 / 2].x);                                 \
+    } while (0)
+
+
+static inline ulong blake2b(uint64_t const nonce, constant ulong* h)
+{
+    ulong2 vv[8] = {
+        {nano_xor_iv0, iv1}, {iv2, iv3},          {iv4, iv5},
+        {iv6, iv7},          {iv0, iv1},          {iv2, iv3},
+        {nano_xor_iv4, iv5}, {nano_xor_iv6, iv7},
+    };
+
+    ROUND(nonce, h[0], h[1], h[2], h[3], 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+    ROUND(0, 0, h[3], 0, 0, 0, 0, 0, h[0], 0, nonce, h[1], 0, 0, 0, h[2]);
+    ROUND(0, 0, 0, nonce, 0, h[1], 0, 0, 0, 0, h[2], 0, 0, h[0], 0, h[3]);
+    ROUND(0, 0, h[2], h[0], 0, 0, 0, 0, h[1], 0, 0, 0, h[3], nonce, 0, 0);
+    ROUND(0, nonce, 0, 0, h[1], h[3], 0, 0, 0, h[0], 0, 0, 0, 0, h[2], 0);
+    ROUND(h[1], 0, 0, 0, nonce, 0, 0, h[2], h[3], 0, 0, 0, 0, 0, h[0], 0);
+    ROUND(0, 0, h[0], 0, 0, 0, h[3], 0, nonce, 0, 0, h[2], 0, h[1], 0, 0);
+    ROUND(0, 0, 0, 0, 0, h[0], h[2], 0, 0, nonce, 0, h[3], 0, 0, h[1], 0);
+    ROUND(0, 0, 0, 0, 0, h[2], nonce, 0, 0, h[1], 0, 0, h[0], h[3], 0, 0);
+    ROUND(0, h[1], 0, h[3], 0, 0, h[0], 0, 0, 0, 0, 0, h[2], 0, 0, nonce);
+    ROUND(nonce, h[0], h[1], h[2], h[3], 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+    ROUND(0, 0, h[3], 0, 0, 0, 0, 0, h[0], 0, nonce, h[1], 0, 0, 0, h[2]);
+
+    return nano_xor_iv0 ^ vv[0].x ^ vv[4].x;
+}
+#undef G32
+#undef G2v
+#undef G2v_split
+#undef ROUND
+
+static inline uint64_t to_u64(constant uint* lo_hi)
+{
+    return static_cast<uint64_t>(uint64_t(lo_hi[1])) << 32UL | lo_hi[0];
+}
+
+static inline void atomic_store_result(volatile device atomic_uint* out, uint64_t val, uint64_t difficulty)
+{
+    uint expected = 0;
+    if (atomic_compare_exchange_weak_explicit(&out[0], &expected, val & 0x00000000ffffffffULL, memory_order_relaxed, memory_order_relaxed))
+    {
+        atomic_store_explicit(&out[1], static_cast<uint>(val >> 32), memory_order_relaxed);
+        atomic_store_explicit(&out[2], static_cast<uint>(difficulty & 0x00000000ffffffffULL), memory_order_relaxed);
+        atomic_store_explicit(&out[3], static_cast<uint>(difficulty >> 32), memory_order_relaxed);
```

Comment:
I don't think these stores need to be atomic, since we've already exclusively atomically set the first segment.

Reply:
Agreed, this was simple copy/paste and it's only done once, so I didn't bother. I will look into it though.

Reply:
Okay, so MSL compiler insists on atomic_* ops on atomic_uint's, so leaving as is for now.